File: Python/dfs_bfs/BJ14500.py
from sys import stdin
input = stdin.readline

def dfs( row, col, length, hap, arr, drow, dcol ):
    global ans
    global n
    global m
    if length == 4:
        ans = max( [ ans, hap ] )
        return
    
    for i in range( 4 ):
        nrow = row + drow[i]
        ncol = col + dcol[i]
        
        if nrow<0 or ncol<0 or nrow>=n or ncol>=m: continue
        if visit[nrow][ncol] == 1: continue
        
        visit[nrow][ncol] = 1
        dfs( nrow, ncol, length+1, hap+arr[row][col], arr, drow, dcol )
        visit[nrow][ncol] = 0

def additionalPlus( row, col, arr, n, m ):
    global ans
    #ㅗ
    hap = 0
    if row-1>=0 and col-1>=0 and col+1<m:
        hap = arr[row][col-1] + arr[row][col] + arr[row][col+1] + arr[row-1][col]
        ans = max( [hap, ans] )
    
    #ㅜ
    hap = 0
    if row+1<n and col-1>=0 and col+1<m:
        hap = arr[row][col-1] + arr[row][col] + arr[row][col+1] + arr[row+1][col]
        ans = max( [hap, ans] )
    
    #ㅏ
    hap = 0
    if row-1>=0 and row+1<n and col+1<m:
        hap = arr[row-1][col] + arr[row][col] + arr[row+1][col] + arr[row][col+1]
        ans = max( [hap, ans] )
    
    #ㅓ
    hap = 0
    if row-1>=0 and row+1<n and col-1<m:
        hap = arr[row-1][col] + arr[row][col] + arr[row+1][col] + arr[row][col-1]
        ans = max( [hap, ans] )
        
if __name__ == "__main__":
    n, m = map(int, input().split())
    drow = [ 1, -1, 0, 0 ]
    dcol = [ 0, 0, 1, -1 ]
    ans = 0
    arr = []
    visit = [ [0]*m for _ in range( n ) ]
    for i in range( n ):
        arr.append( list( map(int, input().split()) ) )
    
    for i in range( n ):
        for j in range( m ):
            visit[i][j] = 1
            dfs( i, j, 0, 0, arr, drow, dcol )
            visit[i][j] = 0
            
            additionalPlus( i, j, arr, n, m )
            
    print( ans )

File: Java/bruteforce/BJ14500.java
package bruteforce;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BJ14500 {
	static int n, m, ans;
	static int[][] arr, visit;
	static int[] dx = { 0, 0, 1, -1 };
	static int[] dy = { 1, -1, 0, 0 };
	
	public static void dfs( int row, int col, int len, int sum ) {
		if( len == 4 ) {
			ans = Math.max( ans, sum );
			return;
		}
		
		for( int i=0; i<4; i++ ) {
			int nrow = row + dy[i];
			int ncol = col + dx[i];
			
			if( nrow<0 || ncol<0 || nrow>=n || ncol>=m ) continue;
			if( visit[nrow][ncol] == 1 ) continue;
			
			visit[nrow][ncol] = 1;
			dfs( nrow, ncol, len+1, sum+arr[nrow][ncol] );
			visit[nrow][ncol] = 0;
		}
	}
	
	public static void additionalPlus( int row, int col ) {
		int sum = 0;
		
		//ㅗ ㅜ ㅏ ㅓ 순으로 비교
		//ㅗ
		if( row-1>=0 && col-1>=0 && col+1<m ) {
			sum = arr[row-1][col] + arr[row][col-1] + arr[row][col] + arr[row][col+1];
			ans = Math.max( ans, sum );
			sum = 0;
		}
		
		//ㅜ
		if( row+1<n && col-1>=0 && col+1<m ) {
			sum = arr[row+1][col] + arr[row][col-1] + arr[row][col] + arr[row][col+1];
			ans = Math.max( ans, sum );
			sum = 0;
		}
		
		//ㅏ
		if( row-1>=0 && row+1<n && col+1<m ) {
			sum = arr[row-1][col] + arr[row][col] + arr[row][col+1] + arr[row+1][col];
			ans = Math.max( ans, sum );
			sum = 0;
		}
		
		//ㅓ
		if( row-1>=0 && row+1<n && col-1>=0 ) {
			sum = arr[row-1][col] + arr[row][col] + arr[row][col-1] + arr[row+1][col];
			ans = Math.max( ans, sum );
			sum = 0;
		}
	}

	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		st = new StringTokenizer( br.readLine() );
		n = Integer.parseInt( st.nextToken() );
		m = Integer.parseInt( st.nextToken() );
		
		arr = new int[n][m];
		visit = new int[n][m];
		for( int i=0; i<n; i++ ) {
			st = new StringTokenizer( br.readLine() );
			for( int j=0; j<m; j++ ) {
				arr[i][j] = Integer.parseInt( st.nextToken() );
			}
		}
		
		for( int i=0; i<n; i++ ) {
			for( int j=0; j<m; j++ ) {
				visit[i][j] = 1;
				dfs( i, j, 1, arr[i][j] );
				visit[i][j] = 0;
				additionalPlus( i, j );
			}
		}
		System.out.println( ans );
		
	}
}